File: agenkit-ts/src/evaluation/metrics.ts
/**
 * Enhanced Metrics Tracking for Agent Evaluation
 *
 * This module extends core evaluation with enhanced metric tracking including:
 * - Session status tracking (running, completed, failed, etc.)
 * - Error collection and analysis
 * - Metric type categorization
 * - Cross-session aggregation
 *
 * Key use case: "How do you know a long-running agent succeeded?"
 *
 * Example:
 * ```typescript
 * const result = new SessionResult('session-123', 'my-agent');
 * result.addMetricMeasurement(
 *   createMetricMeasurement('accuracy', 0.95, MetricType.SuccessRate)
 * );
 * result.setStatus(SessionStatus.Completed);
 *
 * const collector = new MetricsCollector();
 * collector.addSession(result);
 * const aggregate = collector.getAggregatedMetrics();
 * ```
 */

/**
 * Session status types.
 */
export enum SessionStatus {
  /** Session is currently running */
  Running = 'running',
  /** Session completed successfully */
  Completed = 'completed',
  /** Session failed */
  Failed = 'failed',
  /** Session timed out */
  Timeout = 'timeout',
  /** Session was cancelled */
  Cancelled = 'cancelled',
}

/**
 * Metric type categories.
 */
export enum MetricType {
  /** Success/failure rates */
  SuccessRate = 'success_rate',
  /** Output quality scores */
  QualityScore = 'quality_score',
  /** Token/API costs */
  Cost = 'cost',
  /** Time taken */
  Duration = 'duration',
  /** Error frequency */
  ErrorRate = 'error_rate',
  /** Task completion */
  TaskCompletion = 'task_completion',
  /** Custom metrics */
  Custom = 'custom',
}

/**
 * Single metric measurement.
 *
 * Note: This is distinct from the Metric interface in core.
 * Metric interface defines how to measure, MetricMeasurement stores the measurement.
 */
export interface MetricMeasurement {
  /** Name of the metric */
  name: string;
  /** Value of the measurement */
  value: number;
  /** Type categorizes the metric */
  type: MetricType;
  /** Timestamp when measurement was taken */
  timestamp: Date;
  /** Additional context */
  metadata: Record<string, unknown>;
}

/**
 * Create a new metric measurement with current timestamp.
 *
 * @param name Metric name
 * @param value Metric value
 * @param type Metric type
 * @param metadata Optional metadata
 * @returns Metric measurement
 */
export function createMetricMeasurement(
  name: string,
  value: number,
  type: MetricType,
  metadata?: Record<string, unknown>
): MetricMeasurement {
  return {
    name,
    value,
    type,
    timestamp: new Date(),
    metadata: metadata || {},
  };
}

/**
 * Error record for tracking errors during evaluation.
 */
export interface ErrorRecord {
  /** Type of error */
  type: string;
  /** Error message */
  message: string;
  /** Additional details */
  details: Record<string, unknown>;
  /** Timestamp when error occurred */
  timestamp: Date;
}

/**
 * Create a new error record with current timestamp.
 *
 * @param errorType Error type
 * @param message Error message
 * @param details Optional details
 * @returns Error record
 */
export function createErrorRecord(
  errorType: string,
  message: string,
  details?: Record<string, unknown>
): ErrorRecord {
  return {
    type: errorType,
    message,
    details: details || {},
    timestamp: new Date(),
  };
}

/**
 * Session result with enhanced tracking.
 *
 * This extends the core EvaluationResult with session status, error tracking,
 * and richer metadata for long-running agent evaluations.
 */
export class SessionResult {
  sessionId: string;
  agentName: string;
  status: SessionStatus;
  startTime: Date;
  endTime: Date | null = null;
  measurements: MetricMeasurement[] = [];
  errors: ErrorRecord[] = [];
  metadata: Record<string, unknown> = {};

  /**
   * Create session result.
   *
   * @param sessionId Session identifier
   * @param agentName Agent name
   */
  constructor(sessionId: string, agentName: string) {
    this.sessionId = sessionId;
    this.agentName = agentName;
    this.status = SessionStatus.Running;
    this.startTime = new Date();
  }

  /**
   * Set session status.
   *
   * @param status New status
   */
  setStatus(status: SessionStatus): void {
    this.status = status;
    if (
      status !== SessionStatus.Running &&
      status !== SessionStatus.Cancelled &&
      this.endTime === null
    ) {
      this.endTime = new Date();
    }
  }

  /**
   * Add metric measurement.
   *
   * @param measurement Metric measurement
   */
  addMetricMeasurement(measurement: MetricMeasurement): void {
    this.measurements.push(measurement);
  }

  /**
   * Add error record.
   *
   * @param error Error record
   */
  addError(error: ErrorRecord): void {
    this.errors.push(error);
  }

  /**
   * Get duration in seconds.
   *
   * @returns Duration in seconds (0 if not ended)
   */
  getDurationSeconds(): number {
    if (!this.endTime) {
      return 0;
    }
    return (this.endTime.getTime() - this.startTime.getTime()) / 1000;
  }

  /**
   * Get measurements by metric name.
   *
   * @param name Metric name
   * @returns Measurements for this metric
   */
  getMeasurementsByName(name: string): MetricMeasurement[] {
    return this.measurements.filter(m => m.name === name);
  }

  /**
   * Get measurements by metric type.
   *
   * @param type Metric type
   * @returns Measurements of this type
   */
  getMeasurementsByType(type: MetricType): MetricMeasurement[] {
    return this.measurements.filter(m => m.type === type);
  }

  /**
   * Get errors by type.
   *
   * @param type Error type
   * @returns Errors of this type
   */
  getErrorsByType(type: string): ErrorRecord[] {
    return this.errors.filter(e => e.type === type);
  }

  /**
   * Check if session has errors.
   *
   * @returns True if errors occurred
   */
  hasErrors(): boolean {
    return this.errors.length > 0;
  }

  /**
   * Get error count.
   *
   * @returns Number of errors
   */
  getErrorCount(): number {
    return this.errors.length;
  }

  /**
   * Convert to plain object.
   *
   * @returns Plain object representation
   */
  toDict(): Record<string, unknown> {
    return {
      session_id: this.sessionId,
      agent_name: this.agentName,
      status: this.status,
      start_time: this.startTime.toISOString(),
      end_time: this.endTime ? this.endTime.toISOString() : null,
      duration_seconds: this.getDurationSeconds(),
      measurements: this.measurements.map(m => ({
        name: m.name,
        value: m.value,
        type: m.type,
        timestamp: m.timestamp.toISOString(),
        metadata: m.metadata,
      })),
      errors: this.errors.map(e => ({
        type: e.type,
        message: e.message,
        details: e.details,
        timestamp: e.timestamp.toISOString(),
      })),
      metadata: this.metadata,
    };
  }
}

/**
 * Aggregated metric statistics.
 */
export interface AggregatedMetric {
  /** Metric name */
  name: string;
  /** Metric type */
  type: MetricType;
  /** Mean value across sessions */
  mean: number;
  /** Minimum value */
  min: number;
  /** Maximum value */
  max: number;
  /** Standard deviation */
  std: number;
  /** Number of measurements */
  count: number;
}

/**
 * Cross-session metrics collector.
 *
 * Aggregates metrics across multiple evaluation sessions for analysis.
 */
export class MetricsCollector {
  private sessions: Map<string, SessionResult> = new Map();

  /**
   * Add session result.
   *
   * @param session Session result
   */
  addSession(session: SessionResult): void {
    this.sessions.set(session.sessionId, session);
  }

  /**
   * Get session by ID.
   *
   * @param sessionId Session identifier
   * @returns Session result if found
   */
  getSession(sessionId: string): SessionResult | undefined {
    return this.sessions.get(sessionId);
  }

  /**
   * Get all sessions.
   *
   * @returns All session results
   */
  getAllSessions(): SessionResult[] {
    return Array.from(this.sessions.values());
  }

  /**
   * Get sessions by status.
   *
   * @param status Session status
   * @returns Sessions with this status
   */
  getSessionsByStatus(status: SessionStatus): SessionResult[] {
    return this.getAllSessions().filter(s => s.status === status);
  }

  /**
   * Get sessions by agent name.
   *
   * @param agentName Agent name
   * @returns Sessions for this agent
   */
  getSessionsByAgent(agentName: string): SessionResult[] {
    return this.getAllSessions().filter(s => s.agentName === agentName);
  }

  /**
   * Get aggregated metrics across all sessions.
   *
   * @returns Map of metric name to aggregated statistics
   */
  getAggregatedMetrics(): Map<string, AggregatedMetric> {
    const metricsByName = new Map<string, MetricMeasurement[]>();

    // Collect all measurements by metric name
    for (const session of this.sessions.values()) {
      for (const measurement of session.measurements) {
        if (!metricsByName.has(measurement.name)) {
          metricsByName.set(measurement.name, []);
        }
        metricsByName.get(measurement.name)!.push(measurement);
      }
    }

    // Aggregate each metric
    const aggregated = new Map<string, AggregatedMetric>();
    for (const [name, measurements] of metricsByName.entries()) {
      aggregated.set(name, this.aggregateMeasurements(name, measurements));
    }

    return aggregated;
  }

  /**
   * Aggregate measurements for a single metric.
   *
   * @param name Metric name
   * @param measurements Measurements to aggregate
   * @returns Aggregated statistics
   */
  private aggregateMeasurements(
    name: string,
    measurements: MetricMeasurement[]
  ): AggregatedMetric {
    if (measurements.length === 0) {
      return {
        name,
        type: MetricType.Custom,
        mean: 0,
        min: 0,
        max: 0,
        std: 0,
        count: 0,
      };
    }

    const values = measurements.map(m => m.value);
    const sum = values.reduce((a, b) => a + b, 0);
    const mean = sum / values.length;
    const min = Math.min(...values);
    const max = Math.max(...values);

    // Calculate standard deviation
    const variance = values.reduce((acc, val) => acc + Math.pow(val - mean, 2), 0) / values.length;
    const std = Math.sqrt(variance);

    return {
      name,
      type: measurements[0].type,
      mean,
      min,
      max,
      std,
      count: values.length,
    };
  }

  /**
   * Get success rate across all sessions.
   *
   * @returns Success rate (completed / total)
   */
  getSuccessRate(): number {
    const total = this.sessions.size;
    if (total === 0) {
      return 0;
    }

    const completed = this.getSessionsByStatus(SessionStatus.Completed).length;
    return completed / total;
  }

  /**
   * Get error rate across all sessions.
   *
   * @returns Error rate (sessions with errors / total)
   */
  getErrorRate(): number {
    const total = this.sessions.size;
    if (total === 0) {
      return 0;
    }

    const withErrors = this.getAllSessions().filter(s => s.hasErrors()).length;
    return withErrors / total;
  }

  /**
   * Get total error count.
   *
   * @returns Total number of errors
   */
  getTotalErrorCount(): number {
    return this.getAllSessions().reduce((sum, s) => sum + s.getErrorCount(), 0);
  }

  /**
   * Get average duration across completed sessions.
   *
   * @returns Average duration in seconds
   */
  getAverageDuration(): number {
    const completed = this.getSessionsByStatus(SessionStatus.Completed);
    if (completed.length === 0) {
      return 0;
    }

    const totalDuration = completed.reduce((sum, s) => sum + s.getDurationSeconds(), 0);
    return totalDuration / completed.length;
  }

  /**
   * Clear all sessions.
   */
  clear(): void {
    this.sessions.clear();
  }

  /**
   * Get session count.
   *
   * @returns Number of sessions
   */
  getSessionCount(): number {
    return this.sessions.size;
  }
}

File: agenkit-rust/src/evaluation/metrics.rs
//! Enhanced Metrics Tracking
//!
//! This module extends core evaluation with enhanced metric tracking including:
//! - Session status tracking (running, completed, failed, etc.)
//! - Error collection and analysis
//! - Metric type categorization
//! - Cross-session aggregation
//!
//! # Key Use Case
//!
//! "How do you know a long-running agent succeeded?"
//!
//! # Example
//!
//! ```
//! use agenkit::evaluation::{SessionResult, MetricMeasurement, MetricType, SessionStatus};
//!
//! let mut result = SessionResult::new("session-123", "my-agent");
//! result.add_metric_measurement(MetricMeasurement::new(
//!     "accuracy",
//!     0.95,
//!     MetricType::SuccessRate,
//! ));
//! result.set_status(SessionStatus::Completed);
//! ```

use std::collections::HashMap;
use chrono::{DateTime, Utc};
use serde::{Serialize, Deserialize};

/// Status of an evaluation session.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum SessionStatus {
    /// Session is currently running
    Running,
    /// Session completed successfully
    Completed,
    /// Session failed
    Failed,
    /// Session timed out
    Timeout,
    /// Session was cancelled
    Cancelled,
}

impl std::fmt::Display for SessionStatus {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            SessionStatus::Running => write!(f, "running"),
            SessionStatus::Completed => write!(f, "completed"),
            SessionStatus::Failed => write!(f, "failed"),
            SessionStatus::Timeout => write!(f, "timeout"),
            SessionStatus::Cancelled => write!(f, "cancelled"),
        }
    }
}

/// Categorizes different types of metrics.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum MetricType {
    /// Measures success/failure rates
    SuccessRate,
    /// Measures output quality
    QualityScore,
    /// Measures token/API costs
    Cost,
    /// Measures time taken
    Duration,
    /// Measures error frequency
    ErrorRate,
    /// Measures task completion
    TaskCompletion,
    /// Custom metrics
    Custom,
}

impl std::fmt::Display for MetricType {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            MetricType::SuccessRate => write!(f, "success_rate"),
            MetricType::QualityScore => write!(f, "quality_score"),
            MetricType::Cost => write!(f, "cost"),
            MetricType::Duration => write!(f, "duration"),
            MetricType::ErrorRate => write!(f, "error_rate"),
            MetricType::TaskCompletion => write!(f, "task_completion"),
            MetricType::Custom => write!(f, "custom"),
        }
    }
}

/// Single metric measurement.
///
/// Note: This is distinct from the Metric trait in core.rs.
/// Metric trait defines how to measure, MetricMeasurement stores the measurement.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MetricMeasurement {
    /// Name of the metric
    pub name: String,
    /// Value of the measurement
    pub value: f64,
    /// Type categorizes the metric
    #[serde(rename = "type")]
    pub metric_type: MetricType,
    /// Timestamp when measurement was taken
    pub timestamp: DateTime<Utc>,
    /// Metadata for additional context
    #[serde(skip_serializing_if = "HashMap::is_empty")]
    pub metadata: HashMap<String, serde_json::Value>,
}

impl MetricMeasurement {
    /// Creates a new metric measurement with current timestamp.
    pub fn new(name: impl Into<String>, value: f64, metric_type: MetricType) -> Self {
        Self {
            name: name.into(),
            value,
            metric_type,
            timestamp: Utc::now(),
            metadata: HashMap::new(),
        }
    }

    /// Creates a new metric measurement with metadata.
    pub fn with_metadata(
        name: impl Into<String>,
        value: f64,
        metric_type: MetricType,
        metadata: HashMap<String, serde_json::Value>,
    ) -> Self {
        Self {
            name: name.into(),
            value,
            metric_type,
            timestamp: Utc::now(),
            metadata,
        }
    }
}

/// Error that occurred during evaluation.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ErrorRecord {
    /// Type of error
    #[serde(rename = "type")]
    pub error_type: String,
    /// Error message
    pub message: String,
    /// Additional details
    #[serde(skip_serializing_if = "HashMap::is_empty")]
    pub details: HashMap<String, serde_json::Value>,
    /// Timestamp when error occurred
    pub timestamp: DateTime<Utc>,
}

impl ErrorRecord {
    /// Creates a new error record with current timestamp.
    pub fn new(error_type: impl Into<String>, message: impl Into<String>) -> Self {
        Self {
            error_type: error_type.into(),
            message: message.into(),
            details: HashMap::new(),
            timestamp: Utc::now(),
        }
    }

    /// Creates a new error record with details.
    pub fn with_details(
        error_type: impl Into<String>,
        message: impl Into<String>,
        details: HashMap<String, serde_json::Value>,
    ) -> Self {
        Self {
            error_type: error_type.into(),
            message: message.into(),
            details,
            timestamp: Utc::now(),
        }
    }
}

/// Results from evaluating an agent session with enhanced tracking.
///
/// This extends the core EvaluationResult with session status, error tracking,
/// and richer metadata for long-running agent evaluations.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SessionResult {
    /// SessionID uniquely identifies this session
    pub session_id: String,
    /// AgentName identifies the agent being evaluated
    pub agent_name: String,
    /// Status of the session
    pub status: SessionStatus,
    /// StartTime when session started
    pub start_time: DateTime<Utc>,
    /// EndTime when session ended (None if still running)
    pub end_time: Option<DateTime<Utc>>,
    /// Measurements collected during session
    pub measurements: Vec<MetricMeasurement>,
    /// Errors that occurred during session
    pub errors: Vec<ErrorRecord>,
    /// Metadata for additional context
    #[serde(skip_serializing_if = "HashMap::is_empty")]
    pub metadata: HashMap<String, serde_json::Value>,
}

impl SessionResult {
    /// Creates a new session result.
    pub fn new(session_id: impl Into<String>, agent_name: impl Into<String>) -> Self {
        Self {
            session_id: session_id.into(),
            agent_name: agent_name.into(),
            status: SessionStatus::Running,
            start_time: Utc::now(),
            end_time: None,
            measurements: Vec::new(),
            errors: Vec::new(),
            metadata: HashMap::new(),
        }
    }

    /// Sets the session status.
    pub fn set_status(&mut self, status: SessionStatus) {
        self.status = status;
        if self.end_time.is_none() && status != SessionStatus::Running {
            self.end_time = Some(Utc::now());
        }
    }

    /// Adds a metric measurement.
    pub fn add_metric_measurement(&mut self, measurement: MetricMeasurement) {
        self.measurements.push(measurement);
    }

    /// Adds an error record.
    ///
    /// If the session is Running or Completed, it will be marked as Failed.
    pub fn add_error(&mut self, error: ErrorRecord) {
        self.errors.push(error);
        if self.status == SessionStatus::Running || self.status == SessionStatus::Completed {
            self.status = SessionStatus::Failed;
            if self.end_time.is_none() {
                self.end_time = Some(Utc::now());
            }
        }
    }

    /// Gets measurements by metric type.
    pub fn get_measurements_by_type(&self, metric_type: MetricType) -> Vec<&MetricMeasurement> {
        self.measurements
            .iter()
            .filter(|m| m.metric_type == metric_type)
            .collect()
    }

    /// Gets measurements by metric name.
    pub fn get_measurements_by_name(&self, name: &str) -> Vec<&MetricMeasurement> {
        self.measurements
            .iter()
            .filter(|m| m.name == name)
            .collect()
    }

    /// Calculates duration in seconds.
    pub fn duration_secs(&self) -> Option<f64> {
        self.end_time.map(|end| {
            (end - self.start_time).num_milliseconds() as f64 / 1000.0
        })
    }

    /// Checks if session succeeded.
    pub fn is_successful(&self) -> bool {
        self.status == SessionStatus::Completed && self.errors.is_empty()
    }
}

/// Collector for aggregating metrics across multiple sessions.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MetricsCollector {
    /// Sessions tracked by this collector
    pub sessions: Vec<SessionResult>,
    /// Collector metadata
    pub metadata: HashMap<String, serde_json::Value>,
}

impl MetricsCollector {
    /// Creates a new metrics collector.
    pub fn new() -> Self {
        Self {
            sessions: Vec::new(),
            metadata: HashMap::new(),
        }
    }

    /// Adds a session result.
    pub fn add_session(&mut self, session: SessionResult) {
        self.sessions.push(session);
    }

    /// Gets all measurements of a specific type across all sessions.
    pub fn get_all_measurements_by_type(&self, metric_type: MetricType) -> Vec<&MetricMeasurement> {
        self.sessions
            .iter()
            .flat_map(|s| s.get_measurements_by_type(metric_type))
            .collect()
    }

    /// Gets all measurements of a specific name across all sessions.
    pub fn get_all_measurements_by_name(&self, name: &str) -> Vec<&MetricMeasurement> {
        self.sessions
            .iter()
            .flat_map(|s| s.get_measurements_by_name(name))
            .collect()
    }

    /// Calculates aggregated statistics for a metric type.
    pub fn aggregate_by_type(&self, metric_type: MetricType) -> HashMap<String, f64> {
        let measurements = self.get_all_measurements_by_type(metric_type);
        let values: Vec<f64> = measurements.iter().map(|m| m.value).collect();
        Self::calculate_statistics(&values)
    }

    /// Calculates aggregated statistics for a metric name.
    pub fn aggregate_by_name(&self, name: &str) -> HashMap<String, f64> {
        let measurements = self.get_all_measurements_by_name(name);
        let values: Vec<f64> = measurements.iter().map(|m| m.value).collect();
        Self::calculate_statistics(&values)
    }

    /// Calculates basic statistics for a set of values.
    fn calculate_statistics(values: &[f64]) -> HashMap<String, f64> {
        let mut stats = HashMap::new();

        if values.is_empty() {
            return stats;
        }

        let sum: f64 = values.iter().sum();
        let count = values.len() as f64;
        let mean = sum / count;

        stats.insert("mean".to_string(), mean);
        stats.insert("count".to_string(), count);
        stats.insert("sum".to_string(), sum);

        if !values.is_empty() {
            let min = values.iter().fold(f64::INFINITY, |a, &b| a.min(b));
            let max = values.iter().fold(f64::NEG_INFINITY, |a, &b| a.max(b));
            stats.insert("min".to_string(), min);
            stats.insert("max".to_string(), max);

            // Calculate standard deviation
            let variance = values.iter()
                .map(|v| (v - mean).powi(2))
                .sum::<f64>() / count;
            stats.insert("std".to_string(), variance.sqrt());
        }

        stats
    }

    /// Gets success rate across all sessions.
    pub fn overall_success_rate(&self) -> f64 {
        if self.sessions.is_empty() {
            return 0.0;
        }

        let successful = self.sessions.iter().filter(|s| s.is_successful()).count();
        successful as f64 / self.sessions.len() as f64
    }

    /// Gets total error count across all sessions.
    pub fn total_errors(&self) -> usize {
        self.sessions.iter().map(|s| s.errors.len()).sum()
    }
}

impl Default for MetricsCollector {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_session_result_basic() {
        let mut session = SessionResult::new("test-123", "agent-1");

        assert_eq!(session.session_id, "test-123");
        assert_eq!(session.agent_name, "agent-1");
        assert_eq!(session.status, SessionStatus::Running);
        assert!(session.end_time.is_none());

        session.set_status(SessionStatus::Completed);
        assert_eq!(session.status, SessionStatus::Completed);
        assert!(session.end_time.is_some());
    }

    #[test]
    fn test_session_result_measurements() {
        let mut session = SessionResult::new("test-123", "agent-1");

        session.add_metric_measurement(MetricMeasurement::new(
            "accuracy",
            0.95,
            MetricType::SuccessRate,
        ));

        session.add_metric_measurement(MetricMeasurement::new(
            "latency",
            100.0,
            MetricType::Duration,
        ));

        assert_eq!(session.measurements.len(), 2);

        let accuracy_measurements = session.get_measurements_by_name("accuracy");
        assert_eq!(accuracy_measurements.len(), 1);
        assert_eq!(accuracy_measurements[0].value, 0.95);

        let duration_measurements = session.get_measurements_by_type(MetricType::Duration);
        assert_eq!(duration_measurements.len(), 1);
        assert_eq!(duration_measurements[0].value, 100.0);
    }

    #[test]
    fn test_session_result_errors() {
        let mut session = SessionResult::new("test-123", "agent-1");
        session.set_status(SessionStatus::Completed);

        assert!(session.is_successful());

        session.add_error(ErrorRecord::new("timeout", "Request timed out"));

        assert!(!session.is_successful());
        assert_eq!(session.status, SessionStatus::Failed);
        assert_eq!(session.errors.len(), 1);
    }

    #[test]
    fn test_metrics_collector() {
        let mut collector = MetricsCollector::new();

        let mut session1 = SessionResult::new("s1", "agent-1");
        session1.add_metric_measurement(MetricMeasurement::new(
            "accuracy",
            0.9,
            MetricType::SuccessRate,
        ));
        session1.set_status(SessionStatus::Completed);

        let mut session2 = SessionResult::new("s2", "agent-1");
        session2.add_metric_measurement(MetricMeasurement::new(
            "accuracy",
            0.95,
            MetricType::SuccessRate,
        ));
        session2.set_status(SessionStatus::Completed);

        collector.add_session(session1);
        collector.add_session(session2);

        let stats = collector.aggregate_by_name("accuracy");
        assert_eq!(stats.get("count").unwrap(), &2.0);
        assert_eq!(stats.get("mean").unwrap(), &0.925);
        assert_eq!(stats.get("min").unwrap(), &0.9);
        assert_eq!(stats.get("max").unwrap(), &0.95);

        assert_eq!(collector.overall_success_rate(), 1.0);
    }

    #[test]
    fn test_metric_measurement_creation() {
        let measurement = MetricMeasurement::new("test", 0.5, MetricType::QualityScore);

        assert_eq!(measurement.name, "test");
        assert_eq!(measurement.value, 0.5);
        assert_eq!(measurement.metric_type, MetricType::QualityScore);
        assert!(measurement.metadata.is_empty());
    }

    #[test]
    fn test_error_record_creation() {
        let error = ErrorRecord::new("timeout", "Request failed");

        assert_eq!(error.error_type, "timeout");
        assert_eq!(error.message, "Request failed");
        assert!(error.details.is_empty());
    }
}
